File: 2022/Feb/2nd_week/pg_target_number.py

def solution(numbers, target):
    answer = 0

    def dfs(n, num):
        nonlocal answer
        if n == len(numbers):
            if num == target:
                answer += 1
            return
        dfs(n+1, num+numbers[n])
        dfs(n+1, num-numbers[n])

    dfs(0, 0)
    return answer

numbers = [1, 1, 1, 1, 1]
target = 3
target_answer = 5

print(solution(numbers, target))

File: 2022/May/pg_target_number.js
function solution(numbers, target) {
    let answer = 0;

    const recur = (crntValue = 0, n = 0) => {
        if (n == numbers.length) {
            if (target === crntValue) {
                answer += 1
            }
            return
        }

        recur(crntValue + numbers[n], n + 1)
        recur(crntValue - numbers[n], n + 1)
    }

    recur()

    return answer;
}

// case 1
// const numbers = [1, 1, 1, 1, 1]
// const target = 3
// const answer = 5

// case 2
const numbers = [4, 1, 2, 1]
const target = 4
const answer = 2

console.log(solution(numbers, target) === answer)